### DISCUSSION TITLE
CBC와 GCM는 어떠한 차이가 있고 성능에도 차이가 존재할까?

### DISCUSSION THREAD

**Discussion by beatmeJY** [dailyge/dailyge-server#181] [Performance]
## 📝 연관 PR

- [[DLG-88] JWT 내부 필드를 암호화 한다.](https://github.com/dailyge/dailyge-server/pull/50)

<br/><br/>

## ❓ 작성 이유
이전 Discussions에 공유하였던 [JWT 암호화](https://github.com/dailyge/dailyge-server/discussions/122)에서 추가 학습 자료로 남겨두었던 `GCM`에 대해서 동작 원리를 알아보고 `CBC`와 얼마 만큼의 성능 차이도 존재하는지 직접 실험하여 팀원들께 공유하기 위해 작성하였습니다!

<br/><br/>

## 📝 내용
1. GCM의 동작 원리
2. GCM vs CBC 보안적 차이
3. 실제 성능 비교
4. 결론

<br/><br/>

#### `GCM`은 `AES`의 블록 암호화를 병렬로 처리할 수 있는 `CTR`(Counter Mode) 기반 암호화 방식입니다.

<br/>

`CTR`은 각 블록의 독립적인 암호화 처리를 수행하는데,
![image](https://github.com/user-attachments/assets/4031aa94-6b94-45a7-aa9a-66453e2ae018)
위처럼 각 블록을 카운터를 기반으로 병렬로 암호화를 처리할 수 있게 만들어 성능을 크게 향상 시킬 수 있습니다.

<br/>

하지만 `CBC`는 이전 암호화된 블록에 영향을 받기 때문에 병렬 처리가 불가능한데,
![image](https://github.com/user-attachments/assets/641bcdbe-b414-484b-8b8e-fc97e7afdf2b)
각 블록이 이전 블록의 암호화 결과에 의존해` XOR` 암호화 연산을 수행하므로 병렬 처리가 불가능합니다.

<br/><br/>

#### GCM의 동작원리
<img src="https://github.com/user-attachments/assets/450658ef-82cf-4fb8-b952-50f8b40bed11" width="600" height="600" />

**1. IV 생성**
  - IV란 `GCM`에서는 Nonce라고도 불리며 암호화할 때 매번 다른 암호문을 생성하기 위해 사용하는 랜덤 값입니다.
  - GCM에서는 보통 12바이트의 고정된 크기로 생성합니다. (IV가 12바이트가 아닐 경우, 별도의 해시 처리가 필요합니다.)
  - 해당 IV는 암호화할 데이터와 결합되어 카운터 모드로 사용할 초기 값을 만듭니다.

**2. Counter 초기화**
  - CTR 모드를 기반으로 하여 IV와 카운터 값을 결합해 초기 카운터 값을 설정합니다.
  - 카운터는 블록 단위로 증가하며, 이 카운터 값이 `AES` 알고리즘의 입력으로 사용되어 암호화된 출력 블록을 생성합니다.
  - 첫 번째 카운터 블록은 주로 IV(Nonce)를 사용하고, 이후 각 블록에 대해 카운터 값을 1씩 증가시키며 암호화가 진행됩니다.

**3. AES 암호화 수행**
 - 각 블록의 데이터를 대칭키를 사용해 데이터를 암호화합니다.
 - 각 데이터 블록에 대해 카운터 값이 결합된 결과로 암호화 되므로 각 블록은 독립적으로 처리됩니다.

**4. Galois 필드 연산과 인증 태그 생성**
 - 인증 태그인 MAC(Message Authentication Code)은 암호화된 데이터의 무결성을 검증하기 위해 생성됩니다.
 - 인증 태그는 Galois 필드에서 수행되는 수학적 연산에 의해 생성되는데, Galois 필드에서 곱셈 연산을 수행하여 태그 값을 계산합니다.
 - 인증 태그 이외에도 `AAD`(Associated Authenticated Data)가 포함 될 수 있으며 암호화 되지 않지만 인증을 위해 추가적으로 보호됩니다.
 - AAD와 최종적으로 생성된 인증 태그는 암호문과 함께 전송되어, 복호화 시 이 태그를 사용해 데이터가 변조되었는지 확인 할 수 있습니다.

<br/><br/>

## 2. GCM vs CBC 보안적 차이
`CBC`는 Padding이 필요하며 이를 통한 [Padding oracle attack](https://en.wikipedia.org/wiki/Padding_oracle_attack)에 취약한 반면 `GCM`은 `CTR`을 기반으로 각 블록이 독립적으로 동작하기 때문에 블록 암호화이지만 **스트림 암호화** 방식과 유사하게 동작하여 128비트 블록 크기에 맞지 않더라도 블록 크기에 상관없이 필요한 만큼만 데이터를 암호화 할 수 있어 패딩할 필요가 없기 때문에 `CBC`와 같은 패딩 공격에 안전합니다.

또 다른 `GCM`의 특별한 점은 Galois 필드 연산을 추가해 인증 기능을 지원한다는 것입니다.
`GCM`은 `CTR`모드를 기반으로 하여 암호화 할 뿐만 아니라 그 데이터가 조작되지 않았는지 확인하는 인증 태그를 생성하는데,
이 인증 태그는 메시지의 무결성을 보장하며, 중간에서 데이터가 변조되는 경우 탐지할 수 있습니다.

이러한 특성 덕분에 `AEAD`로 분류되며 GCM은 기본적으로 AEAD를 지원해 보안성과 성능이 모두 뛰어난 방식으로 인정받고 많이 사용되고 있습니다.
 - `AEAD`(Authenticated Encryption with Associated Data)란 암호화와 인증을 동시에 진행하고 인증 강화를 위해 관련된 데이터를 추가로 붙인다는 의미를 나타냅니다.

<br/><br/>

## 3. 실제 성능 비교
```java
private static final long REPEAT_COUNT = 1_000_000L;

@Test
@DisplayName("GCM 성능을 테스트한다.")
void whenGcmPerformance() {
    long gcmStartTime = System.nanoTime();

    for (long i = 0; i < REPEAT_COUNT; i++) {
        String encryptedId = gcmDecryptHelper.encryptId(i);
        Long decryptedId = gcmDecryptHelper.decryptId(encryptedId);
        assertEquals(i, decryptedId);
    }

    long gcmEndTime = System.nanoTime();
    long gcmDurationInMillis = TimeUnit.NANOSECONDS.toMillis(gcmEndTime - gcmStartTime);
    System.out.println("GCM Mode Total Execution Time: " + gcmDurationInMillis + " ms");
}

@Test
@DisplayName("CBC 성능을 테스트한다.")
void testCbcPerformance() {
    ... (나머지 동일)
        String encryptedId = cbcDecryptHelper.encryptId(i);
        Long decryptedId = cbcDecryptHelper.decryptId(encryptedId);
    ...
}
```
### GCM 100만 번 실행
![image](https://github.com/user-attachments/assets/cf38a4ce-36cc-4515-b27e-257b3c9ec03e)
<br />
### CBC 100만 번 실행
![image](https://github.com/user-attachments/assets/259fca70-fb44-4736-a97f-2d6c896b3518)

### 성능 테스트 비교 결과
1. 적은 횟수로 반복하여 비교를 할 때에는 거의 성능의 차이가 없거나 들쑥날쑥하게 결과가 나왔지만,
7. 소요 시간이 초 단위 이상으로 나오도록(100만 번 이상) 많이 실행하게 되면 확실하게 약 **15~20%까지 성능의 차이**가 발생했습니다. 

<br/><br/>

## 4. 결론
1. `CBC`는 오래전부터 사용되어 검증이 잘 되어있으며, 참조할 레퍼런스들도 많아 처음 접하기 용이해 사용하였지만,
2. `GCM`은 `CTR`방식으로 진행되어 `Padding`이 필요하지 않아 CBC의 Padding oracle attack과 같은 보안적 취약점이 없어 **보안적**으로 **우세**합니다.
3. 또한 `GCM`의 `CTR` 방식은 병렬 처리가 가능하기 때문에 현대에 병렬처리가 가능한 CPU 아키텍처 덕분에 **성능적**으로도 **우세**하여 개선의 여지가 있다고 생각하였습니다.

때문에 현재 프로젝트에서 JWT 암호화에 사용 중인 `CBC`대신 `GCM` 암호화 알고리즘으로 변경하여 보도록 하겠습니다!

<br/><br/>

## 📌 참조
 - [패딩을 사용하는 CBC 모드 대칭 암호 해독의 타이밍 취약성](https://learn.microsoft.com/ko-kr/dotnet/standard/security/vulnerabilities-cbc-mode)
 - [TLS/암호 알고리즘 쉽게 이해하기 - MAC, AE, AEAD](https://blog.humminglab.io/posts/tls-cryptography-13-mac-aead/#aead-authenticated-encryption-with-associated-data)

**Comment by devjun10** [comment]
이런것도 있었군요. 하나 배워가네요. 👍
